spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -868,42 +874,67 @@ struct ProductionVisitor
-                                 lp.location());
+                    auto switch_ = builder()->addSwitch(builder::id("rc"), location);
 
-            auto switch_ = builder()->addSwitch(builder::id("rc"), lp.location());
+                    // No match, try again.
+                    pushBuilder(switch_.addCase(builder::integer(-1)), [&]() {
+                        auto ok = builder()->addIf(pb->waitForInputOrEod());
+                        ok->addContinue();
+                        builder()->addAssign(state().lahead, look_ahead::Eod);
+                        builder()->addAssign(state().lahead_end, builder::begin(state().cur));
+                        builder()->addBreak();
+                    });
 
-            auto no_match_try_again = switch_.addCase(builder::integer(-1));
-            pushBuilder(no_match_try_again);
-            auto ok = builder()->addIf(pb->waitForInputOrEod());
-            ok->addContinue();
-            builder()->addAssign(state().lahead, look_ahead::Eod);
-            builder()->addAssign(state().lahead_end, builder::begin(state().cur));
-            builder()->addBreak();
-            popBuilder();
+                    // No match, error.
+                    pushBuilder(switch_.addCase(builder::integer(0)), [&]() {
+                        pb->state().printDebug(builder());
+                        builder()->addAssign(state().lahead, look_ahead::None);
+                        builder()->addAssign(state().lahead_end, builder::begin(state().cur));
+                        builder()->addBreak();
+                    });
 
-            auto no_match_error = switch_.addCase(builder::integer(0));
-            pushBuilder(no_match_error);
-            builder()->addAssign(state().lahead, look_ahead::None);
-            builder()->addAssign(state().lahead_end, builder::begin(state().cur));
-            builder()->addBreak();
-            popBuilder();
+                    pushBuilder(switch_.addDefault(), [&]() {
+                        builder()->addAssign(state().lahead, builder::id("rc"));
+                        builder()->addAssign(state().lahead_end, builder::begin(builder::id("ncur")));
+                        builder()->addBreak();
+                    });
+                });
 
-            auto match = switch_.addDefault();
-            pushBuilder(match);
-            builder()->addAssign(state().lahead, builder::id("rc"));
-            builder()->addAssign(state().lahead_end, builder::begin(builder::id("ncur")));
-            builder()->addBreak();
-            popBuilder();
+                pb->state().printDebug(builder());
+            };
+
+            switch ( mode ) {
+                case LiteralMode::Default:
+                case LiteralMode::Try: {
+                    incremental_matching();
+                    break;
+                }
+
+                case LiteralMode::Search: {
+                    // Create a loop for search mode.
```

Comment:
This loop needs to wrap both regexps and literals so we can search on all of them simultaneously.